### ISSUE TITLE
Need test to make sure buildah doesn't list "-dev" in its output of podman info

### ISSUE THREAD

**Issue by TomSweeneyRedHat** [containers/podman#5278]
Add a test to make sure there is never the letters "-dev" in the podman info | grep -i buildah before we release a new podman. Don't want that gating PRs but only releases.

**Comment by TomSweeneyRedHat** [comment]
From #5253 comments

**Comment by edsantiago** [comment]
What difference is there in test environment between regular PR and release-gating? That is, how does the CI environment know "this is a release candidate"?

**Comment by rhatdan** [comment]
Good question.  I guess it would be perfect if we could trigger that before pushing a release.

**Comment by rhatdan** [comment]
Well we hit this, this past week.

**Comment by edsantiago** [comment]
@cevich can you suggest a way to differentiate, within a CI run, between release candidates and regular builds?

**Comment by cevich** [comment]
Yes, but it is (unfortunately) less than reliable because of ambiguity in both the definition of a "release" and inconsistent human behaviors WRT order of push operations and commits.  

Ref: https://cirrus-ci.org/guide/writing-tasks/#environment-variables

* The env. var. `$CIRRUS_TAG` will be set if the currently executing run was triggered because someone pushed a new or updated tag.  The `$CIRRUS_BRANCH` will reflect some random branch name if the tagged commit is on more than one.
* If the build was triggered by a push to a branch, then `$CIRRUS_PR` will be empty string and `$CIRRUS_BRANCH` will be the name of the branch.  `$CIRRUS_TAG` will be empty.

However, For the purpose of switching testing behavior on/off, it **might** be mostly okay to just check if `$CIRRUS_TAG` is set or not.   You can observe builds happening against tags, by using the special URL:
``https://cirrus-ci.com/github/containers/libpod/<tag name>``.

To my knowledge, that's the only place they show up.  If those builds fail, I believe they will post the typical 'post-merge failure' message on IRC.  That could be differentiated by adding a conditional on `$CIRRUS_TAG` being empty or not.  See `contrib/cirrus/notice_branch_failure.sh` and you'll find some more ideas in `contrib/cirrus/success.sh` (limited to PRs by 'only_if' on `success` task in `.cirrus.yml`).

**Comment by rhatdan** [comment]
@cevich Any way we can move forward on this?

**Comment by rhatdan** [comment]
@cevich Friendly Ping.

**Comment by cevich** [comment]
My current thinking on this:  [Use a 'manual' task](https://cirrus-ci.org/guide/writing-tasks/#manual-tasks) that's limited to PRs, and an automatic task for branches: Both dependent upon the `success` task.  This allows:

1. A human can run the manual task if they know it's a release PR (a minority of all PRs).  This task has an `only_if` condition on `$CIRRUS_PR` being non-empty.  When it runs, it gives simple go/no-go merge feedback in the PR.

2. Post-merge automation (runs the same script as above) is also dependent on `success` being green.  However, it has an `only_if` condition on `$CIRRUS_PR` and `$CIRRUS_TAG` both being empty (i.e. it only runs for branches).

A key requirement this method satisfies is: It allows humans to use the *re-run* button to get past any flakes/failures.  It also cooperates with having (some kind of) notifications on branch-failure (another TODO of mine).

Make sense or have I missed something?

**Comment by rhatdan** [comment]
SGTM

**Comment by cevich** [comment]
Okay, I'll work this into https://github.com/containers/podman/pull/6784

**Comment by rhatdan** [comment]
@cevich Any progress?

**Comment by cevich** [comment]
> @cevich Any progress?

Thanks for checking, yes in fact I added this really stupid-simple "test":
https://github.com/containers/podman/blob/master/contrib/cirrus/runner.sh#L197-L200

That goes with these two conditional tasks:
https://github.com/containers/podman/blob/master/.cirrus.yml#L618-L655

Unfortunately, neither is well tested...TBH, I only verified that they fail (as expected) :grin:  Worse, and less obvious, the main 'Verify Release' task won't be easily seen if it fails, since we have no notifications setup for tag-push triggered tests :confounded: 

Before I get carried away making extra work for myself... Do you know how I (in a test PR) make the `-dev` "go away" from `podman info` so I can confirm at least the manual check is working?

Gosh I probably should really put all this stuff into a jira card...

**Comment by rhatdan** [comment]
@edsantiago @cevich I thought I saw a test for this happening recently?

**Comment by rhatdan** [comment]
We at least hit this before we release correct?  Even it it requires human intervention?

**Comment by zhangguanzhang** [comment]
related https://github.com/containers/podman/issues/8802

**Comment by cevich** [comment]
@rhatdan @TomSweeneyRedHat @zhangguanzhang We have both a manual and automatic test for this already setup:

* [Task definition](https://github.com/containers/podman/blob/b5028541cfacc1ee344815c047df70d74fde7b24/.cirrus.yml#L654-L692)
* https://github.com/containers/podman/blob/b5028541cfacc1ee344815c047df70d74fde7b24/contrib/cirrus/runner.sh#L201-L205

My thinking was, before release time someone would push a PR that removes the `-dev` tag, and use the manual job to confirm.  Then, once the release PR is pushed, and tagged, the automatic task would run to confirm.  ***However***, this was all implemented before I was aware/understood out [broken release process and failing tag testing job](https://github.com/containers/podman/issues/8538).

In other words:

1. Verifying `-dev` absence is meaningless because all release-time test failures are ignored.  ([doc](https://docs.google.com/document/d/1Nfq8R04XtpFEweExs31Qv6iZVnz-A5BQkOyEzXP9eb8))
2. There is no point during the release process where a release-testing failure can be corrected without pushing another release.
 
So no amount of testing for `-dev` will be effective unless we revise both our release process, and (most importantly) make automated-testing success mandatory for release (somehow).  I have some ideas on this topic, and am interested in helping solve the problems, but it's going to require process changes and teamwork.

**Comment by zhangguanzhang** [comment]
Maybe you can add a judgment, if the file does not contain -dev, the output of judgment info should not be included

**Comment by zhangguanzhang** [comment]
@cevich Maybe you can add a judgment, if the file does not contain -dev, the output of judgment info should not be included

**Comment by cevich** [comment]
> Maybe you can add a judgment

Any number of things could be done if all results were not ignored (and difficult to observe tag-push triggered jobs) during release :confounded: 

Despite that, I think presence/absence of `-dev` has consequences in all cases.  i.e. We also don't want development releases to *not* have `-dev` in the version output.

**Comment by cevich** [comment]
Idea:

Given the podman release process complexities and broken tag-push testing, I think the simplest approach is to just have manual tests, and make executing them part of the release process.

Could that be an acceptable solution?

**Comment by cevich** [comment]
Alternate idea:

Change the process such that on tag-push:

* CI Only runs tasks which build the release, swagger docs, and release-testing job
* Passing CI is required for a newly pushed tag to become an official release

@mheon need your input here.  IMHO it's nonsensical to have automatic release-testing while also ignoring the results (as a matter of tradition and future practice).  However, I also think that the automated release testing needs to come from scripts stored outside the podman repository - this enables re-running a failed release-test after fixing the scripts (when necessary), without also requiring a new tag.

**Comment by mheon** [comment]
Potential problem: is there much distinction between "pushed tag" and "official release"? There are steps after the tag is pushed, but they're mostly docs updates - I think most distros are watching for tag push, so they'll grab the new release even without them.

**Comment by cevich** [comment]
> is there much distinction between "pushed tag" and "official release"?

That's up to us to define.  The most flexible definition for "official release" is: HEAD of a release branch.  The most strict definition is "pushed tag".  However, the built-in definition from the perspective of github is "tag push".  i.e. [the release page](https://github.com/containers/podman/releases) ***only*** supports tags.

@edsantiago asked:

> @cevich can you suggest a way to differentiate, within a CI run, between release candidates and regular builds?

We're limited here to only the operators supported by CI ([Cirrus-CI](https://cirrus-ci.org/guide/writing-tasks/#supported-operators), [GitHub-Actions](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjob_idif), etc.).  For Cirrus-CI, keep in mind this touches the same `.cirrus.yml` attributes used by the (convoluted) `[CI:DOCS]` feature.  i.e. Any differentiating necessarily will make that logic _much_ more complicated / less readable :cry:  Otherwise (ignoring that complication for now) we could/can do things like this:

```yaml
some_task:
    only_if: $CIRRUS_TAG =~ '.+-rc'

other_task:
    only_if: $CIRRUS_TAG !=~ '.+-rc'
```

**Comment by edsantiago** [comment]
>We're limited here to only the operators supported by CI (Cirrus-CI, GitHub-Actions, etc.).

That, I think, is the key: we don't necessarily have to use Cirrus, we could look into Github actions. (Left as an exercise for a reader who has figured those out; I've tried twice, given up both times).

**Comment by cevich** [comment]
> That, I think, is the key: we don't necessarily have to use Cirrus

I've explored Github Actions quite extensively.  In fact, the `cirrus-ci_retrospective` work I did was largely intended as support for release-automation.  Enabling a github-actions workflow to trigger only _after_ Cirrus-CI has completed.  

> I've tried twice, given up both times

Now, while Github Actions provide a lot of conveniences for GitHub interactions, the implementation is quite immature, and Github's oddball security protections require lots of hoop-jumping to get (otherwise simple) things done.

Dealing with tags is particularly annoying, as it's not exposed as a simple value to test against.  Last I looked, the ***only*** "context" variables which contain the tag are [`env.GITHUB_REF`](https://docs.github.com/en/free-pro-team@latest/actions/reference/environment-variables#default-environment-variables) and [`github.ref`](https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions#github-context).  It's value uses multiple formats depending on how the job is run, and there are [no search functions which support wildcards](https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions#functions) of any sort.

The *only* place where it's possible to use a wildcard with a tag, is as [part of the overall workflow trigger.](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#onpushpull_requestbranchestags).  This type of trigger also/only/ever runs using the workflow-YAML file stored on the master branch...i.e. it's a major PITA to test/debug/maintain.  So even using Github Actions requires some significant convoluted/complex/difficult compromises :disappointed:  All that said, we could have the release testing/processing script handle the "only run for a tag-push" condition.

Yet another automation and release-process option is to go the "GitOps" route.  Establish another repository (i.e. `podman_release`), and use it's content and automation to "do the right things" when it comes to podman releases.  For example, this other repo. could have a file which contains the latest "to be released" podman commit.  Automation in the other repo, detects changes to this file, and drives the whole process, accessing/changing the podman repo. contents (tags/branchs) as required.  This could solve a lot of the triggering and operational complexity issues, while keeping the automation config. and scripts separate, supports testing of the release-process itself, and changing the code w/o requiring a new [podman] tag to be pushed.  Sorry...I know this approach is a bit "headdy"...but it's not unprecedented.

**Comment by cevich** [comment]
:sweat: 

Taking a step back here...this specific issue "technically" [is resolved by my PR](https://github.com/containers/podman/pull/8900#issuecomment-755404822).  Further, the (single) release test ***can*** be exercised within a PR by pressing the "Optional Release Test" trigger button, [for example](https://cirrus-ci.com/task/6339722370875392).  However as I stated above, no amount of testing matters if the results are ignored.  So that's a key human element we must address (including it's cause).

**Comment by edsantiago** [comment]
I understood maybe three words there. IIUC it sounds like #8900 is your recommendation for the best way to go, and I'm OK with that.

**Comment by mheon** [comment]
@cevich If the goal of this is to prevent full releases from happening without a `-dev` suffix, anything that happens after the tag is pushed is too late - we may not consider the release "official", but unless it was suffixed `-rc`, packages are auto-building for every major distribution including the bad commit.

I am unsure of how much effort we should be throwing at this, though. Most releases of Podman are closely preceded by us vendoring an appropriate Buildah release, so this should be manually verified before we even begin the release process. Do we have evidence that we've been forgetting to do this?

**Comment by cevich** [comment]
> Do we have evidence that we've been forgetting to do this?

I'm basing the need for more trust in [improved] release-automation on a statement you made: "We almost always find some major problem in .0 versions right after they go out, and need to quickly follow-up with a .1" (or something to that effect).  Secondarily, my (and most other) arguments pushing for automated releases is to remove the "human" element from what are often lengthy and complex release and publishing work-flows.

> IIUC it sounds like #8900 is your recommendation for the best way to go

Yes, it reduces activities during the run for a tag-push down to the absolute minimum: Build the release binaries/tarballs/zips + run one simple check (for `-dev` absence).  It assumes the entire tree up-to the tagging, has already 100% passed automated (and possibly manual) testing.  Obviously, the tag-push task set is imperfect/incomplete, and could still fail.  But the results certainly should not be ignored as a matter of procedure.

---

All that said, I think there are significant potential benefits to improving the podman (and buildah and ...) release automation.  It's not going to catch every `.0` foible, but it certainly would help avoid those covered by tests.  Anyway, this needs to be part of a larger group-discussion, I'm confident there's a middle-ground solution that would be acceptable to everyone involved.

**Comment by rhatdan** [comment]
@cevich Any update on this?

**Comment by cevich** [comment]
Oh, actually I think we can close it.  We have a working test for this, and I'm [massaging the release-process docs](https://github.com/containers/podman/pull/9142) to help make sure test failure is not ignored for a release.
